rfd/0141-tsh-bench-db.md

```diff
@@ -0,0 +1,279 @@
+
+```code
+tsh bench postgres [--db-user=] [--db-name=] [--uri] [database]
+```
+
+```code
+tsh bench mysql [--db-user=] [--db-name=] [--uri] [database]
+```
+
+| Flag        | Description |
+| ----------- | ----------- |
+| `--db-user` | Database user used to connect to the target database. The user must have enough permissions on the database to execute all the benchmark queries. |
+| `--db-name` | Database name where benchmark queries will be executed. Depending on the database protocol, this isn't required. |
+| `--uri`     | Direct database access URI. When provided, the benchmark will issue connections directly to this database, and no Teleport is involved in the testing. It must contain all the connection information, including authentication credentials. The contents from --db-user and --db-name will be ignored when provided. |
+| `database`  | Teleport target database name. Available databases can be retrieved by running `tsh db ls`. If not provided, the `--uri` flag (for direct database connection) must be present. | 
+
+Note: The `--rate` common flag will indicate how many connections per second will be
+established.
+
+**Examples:**
+```bash
+# Executes 25 connections per second during 10 seconds without errors but also
+# without meeting the target requests.
+$ tsh bench postgres --rate=25 --duration=10s --db-user=postgres --db-name=postgres postgres-dev
+
+* Requests originated: 88
+* Requests failed: 0
+
+Histogram
+
+Percentile Response Duration
+---------- -----------------
+25         1346 ms
+50         3535 ms
+75         5259 ms
+90         5847 ms
+95         6663 ms
+99         7271 ms
+100        7311 ms
+
+# Executes 100 connections per second during 30 seconds with errors (only the
+# last error is shown).
+$ tsh bench postgres --rate 100 --duration 30s --db-user=postgres --db-name=postgres postgres-dev
+
+* Requests originated: 1586
+* Requests failed: 30
+* Last error: failed to connect to `host=localhost user=postgres database=`: dial error (dial tcp [::1]:5433: connect: connection refused)
+
+Histogram
+
+Percentile Response Duration
+---------- -----------------
+25         3773 ms
+50         4639 ms
+75         5263 ms
+90         6243 ms
+95         6775 ms
+99         10567 ms
+100        22783 ms
+```
+
+### Direct database connections
+Directly targeting databases (not through Teleport) can be used by developers
+quickly to generate baseline measures that can be used to determine how much
+overhead Teleport adds compared to raw database connections.
+
+To avoid adding multiple flags to cover the scenarios, there will be a single
+`--uri` flag, which will receive the database URI with all options included.
+This way, users can choose the authentication method that will be used.
+
+By providing this flag, the `tsh` will connect directly to the database without
+using any Teleport certificate.
+
+**Example:**
+```code
+$ tsh bench postgres --uri "postgres://hello@localhost:5432"
```

Comment:
Like mentioned above, I propose we just teach the command to do `tsh bench postgres postgres://hello@xxx` and get rid of the extra flag.